File: frontend/src/lib/api/household.ts
import "server-only";

import { apiFetch } from "@/lib/api/client";
import { requireSessionCookieHeader } from "@/lib/api/session";
import type {
  ClaimSuggestionData,
  HouseholdOverviewData,
} from "@/lib/api/types";
import type { SpouseLink } from "@/lib/types/household";

export async function fetchHouseholdOverview(taxYear?: number) {
  const cookie = await requireSessionCookieHeader();
  const query = taxYear ? `?tax_year=${taxYear}` : "";
  return apiFetch<HouseholdOverviewData>(`/api/v1/household${query}`, { cookie });
}

export async function getHouseholdOverview(taxYear: number) {
  const { body } = await fetchHouseholdOverview(taxYear);
  if (!body.success) {
    throw new Error(body.message);
  }
  return body.data;
}

export async function requestSpouseLinkWithFastApi(payload: {
  partner_email: string;
}) {
  const cookie = await requireSessionCookieHeader();
  return apiFetch<{ link_id: string; status: string }>(
    "/api/v1/household/spouse-link",
    {
      method: "POST",
      body: payload,
      cookie,
    },
  );
}

export async function respondSpouseLinkWithFastApi(
  linkId: string,
  payload: { action: "accept" | "reject" },
) {
  const cookie = await requireSessionCookieHeader();
  return apiFetch<{ link_id: string; status: string }>(
    `/api/v1/household/spouse-link/${linkId}/respond`,
    {
      method: "POST",
      body: payload,
      cookie,
    },
  );
}

export async function dissolveSpouseLinkWithFastApi(linkId: string) {
  const cookie = await requireSessionCookieHeader();
  return apiFetch<null>(`/api/v1/household/spouse-link/${linkId}`, {
    method: "DELETE",
    cookie,
  });
}

export async function reassignReceiptWithFastApi(
  receiptId: string,
  payload: { target_user_id: string },
) {
  const cookie = await requireSessionCookieHeader();
  return apiFetch<null>(
    `/api/v1/household/receipts/${receiptId}/reassign`,
    {
      method: "POST",
      body: payload,
      cookie,
    },
  );
}

export async function fetchClaimSuggestion(receiptId: string) {
  const cookie = await requireSessionCookieHeader();
  return apiFetch<ClaimSuggestionData>(
    `/api/v1/household/receipts/${receiptId}/claim-suggestion`,
    { cookie },
  );
}

export async function getClaimSuggestion(receiptId: string) {
  const { body } = await fetchClaimSuggestion(receiptId);
  if (!body.success) {
    throw new Error(body.message);
  }
  return body.data;
}

export async function requestSpouseLink(partnerEmail: string): Promise<SpouseLink> {
  const { body } = await requestSpouseLinkWithFastApi({ partner_email: partnerEmail });
  if (!body.success) {
    throw new Error(body.message);
  }
  return {
    id: body.data.link_id,
    requester_id: "",
    partner_id: null,
    partner_email: partnerEmail,
    status: body.data.status as SpouseLink["status"],
    created_at: new Date().toISOString(),
    responded_at: null,
  };
}

export async function respondToLink(
  linkId: string,
  action: "accept" | "reject",
): Promise<SpouseLink> {
  const { body } = await respondSpouseLinkWithFastApi(linkId, { action });
  if (!body.success) {
    throw new Error(body.message);
  }
  return {
    id: body.data.link_id,
    requester_id: "",
    partner_id: null,
    partner_email: "",
    status: body.data.status as SpouseLink["status"],
    created_at: new Date().toISOString(),
    responded_at: new Date().toISOString(),
  };
}

export async function dissolveLink(linkId: string): Promise<void> {
  const { body } = await dissolveSpouseLinkWithFastApi(linkId);
  if (!body.success) {
    throw new Error(body.message);
  }
}

export const getHouseholdOverviewFromApi = getHouseholdOverview;

File: backend/app/api/v1/routes/household.py
import uuid

from fastapi import APIRouter, Depends, Query, status
from fastapi.responses import Response
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.deps import CurrentUserDep, OrgAdminDep, get_db
from app.schemas.common import ApiResponse
from app.schemas.household import (
    ClaimSuggestionData,
    HouseholdOverviewData,
    ReceiptReassignRequest,
    SpouseLinkRequest,
    SpouseLinkRespondRequest,
)
from app.services.household import HouseholdService

router = APIRouter(prefix="/household", tags=["household"])


def _household_service(db: AsyncSession = Depends(get_db)) -> HouseholdService:
    return HouseholdService(db)


@router.get("", response_model=ApiResponse[HouseholdOverviewData])
async def get_household_overview(
    current_user: CurrentUserDep,
    service: HouseholdService = Depends(_household_service),
) -> ApiResponse[HouseholdOverviewData]:
    data = await service.get_overview(current_user)
    return ApiResponse(success=True, data=data, message=None)


@router.post("/spouse-link", response_model=ApiResponse[dict], status_code=status.HTTP_201_CREATED)
async def request_spouse_link(
    payload: SpouseLinkRequest,
    current_user: CurrentUserDep,
    service: HouseholdService = Depends(_household_service),
) -> ApiResponse[dict]:
    link = await service.request_spouse_link(current_user, payload)
    return ApiResponse(
        success=True,
        data={"link_id": str(link.id), "status": link.status},
        message=None,
    )


@router.post("/spouse-link/{link_id}/respond", response_model=ApiResponse[dict])
async def respond_spouse_link(
    link_id: uuid.UUID,
    payload: SpouseLinkRespondRequest,
    current_user: CurrentUserDep,
    service: HouseholdService = Depends(_household_service),
) -> ApiResponse[dict]:
    link = await service.respond_to_link(current_user, link_id, payload)
    return ApiResponse(
        success=True,
        data={"link_id": str(link.id), "status": link.status},
        message=None,
    )


@router.delete("/spouse-link/{link_id}", response_model=ApiResponse[None])
async def dissolve_spouse_link(
    link_id: uuid.UUID,
    current_user: CurrentUserDep,
    service: HouseholdService = Depends(_household_service),
) -> ApiResponse[None]:
    await service.dissolve_link(current_user, link_id)
    return ApiResponse(success=True, data=None, message="Pautan pasangan diputuskan.")


@router.post("/receipts/{receipt_id}/reassign", response_model=ApiResponse[None])
async def reassign_receipt_to_spouse(
    receipt_id: uuid.UUID,
    payload: ReceiptReassignRequest,
    current_user: CurrentUserDep,
    service: HouseholdService = Depends(_household_service),
) -> ApiResponse[None]:
    await service.reassign_receipt(
        current_user,
        receipt_id,
        target_user_id=payload.target_user_id,
    )
    return ApiResponse(success=True, data=None, message="Resit dipindahkan.")


@router.get(
    "/receipts/{receipt_id}/claim-suggestion",
    response_model=ApiResponse[ClaimSuggestionData],
)
async def get_claim_suggestion(
    receipt_id: uuid.UUID,
    current_user: CurrentUserDep,
    service: HouseholdService = Depends(_household_service),
) -> ApiResponse[ClaimSuggestionData]:
    data = await service.suggest_claim_owner(current_user, receipt_id)
    return ApiResponse(success=True, data=data, message=None)
